optimum/tpu/distributed_model.py

```diff
@@ -80,9 +90,12 @@ def _mp_fn(rank, model_id, root_mailbox: RootMailbox, sample_fn: callable):
     )
 
     # Model loading and sharding should happen here
-    model = TpuModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float32)
+    config = AutoConfig.from_pretrained(model_id)
+    model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=config.torch_dtype)
```

Comment:
I have hard time to get why we need to do this way? We are overriding the default behaviour to the default behaviour no? @regisss do you know?

Reply:
It seems the default is to load in fp32 whatever the dtype specified in the config is: https://huggingface.slack.com/archives/C014N4749J9/p1712757959601599

Reply:
So I got some insights on the design for this. It seems that transformers uses the default pytorch type, i.e.: `torch.float32`. So probably I will need to change this code later, as it might not work if there are models whose weights were not trained in float32/bfloat16. I have seen we cannot use bf16 everywhere already, because some operations cannot be made (I've seen it in a unit test with gpt2). It is probably a custom configuration we need to add to the model. I pushed a fix cleaner than this.